lib/internal/modules/esm/fetch.js

```diff
@@ -0,0 +1,310 @@
+          } else {
+            onError = function onError(e) {
+              dispose();
+              reject(e);
+              r(e);
+            };
+          }
+          bodyStream.on('error', onError);
+          bodyStream.on('data', (d) => {
+            ArrayPrototypePush(buffers, d);
+            size += d.length;
+          });
+          bodyStream.on('end', () => {
+            const body = entry.body = /** @type {Buffer} */(
+              BufferConcat(buffers, size)
+            );
+            f(body);
+          });
+        }),
+      };
+      cacheForGET.set(parsed.href, entry);
+      fulfill(entry);
+    });
+  });
+  cacheForGET.set(parsed.href, result);
+  return result;
+}
+
+/**
+ * Returns if an address has local status by if it is going to a local
+ * interface or is an address resolved by DNS to be a local interface
+ * @param {string} hostname url.hostname to test
+ * @returns {Promise<boolean>}
+ */
+async function isLocalAddress(hostname) {
+  try {
+    if (StringPrototypeStartsWith(hostname, '[') &&
+        StringPrototypeEndsWith(hostname, ']')) {
+      hostname = StringPrototypeSlice(hostname, 1, -1);
+    }
+    const addr = await dnsLookup(hostname, { verbatim: true });
+    if (addr.family === 4) {
+      return StringPrototypeStartsWith(addr.address, '127.');
+    } else if (addr.family === 6) {
+      return normalizeIPv6(addr.address) ===
+        '0000:0000:0000:0000:0000:0000:0000:0001';
+    }
+  } catch {}
+  return false;
+}
+
+/**
+ * Used to expand interface addresses to a canonical form for comparisons
+ * E.G. comparing ::1, 0::1, etc. to 0000:0000:0000:0000:0000:0000:0000:0001
+ * Due to IPv6 having multiple ways to represent a single address we need to
+ * normalize
+ * @param {string} str
+ * @returns {string}
+ */
+function normalizeIPv6(str) {
+  const parts = StringPrototypeSplit(str, ':');
+  let expansionIndex = -1;
+  for (let i = 0; i < parts.length; i++) {
+    const part = parts[i];
+    if (part === '' && expansionIndex === -1 && parts.length < 8) {
+      expansionIndex = i;
+      continue;
+    }
+    parts[i] = StringPrototypePadStart(part, 4, '0');
+  }
+  if (parts.length < 8) {
+    ArrayPrototypeSplice(
+      parts,
+      expansionIndex,
+      1,
+      ...ArrayFrom({ length: 8 - parts.length + 1 }, () => '0000')
+    );
+  }
+  return ArrayPrototypeJoin(parts, ':');
+}
+
+/**
+ * Fetches a location with a shared cache following redirects.
+ * Does not respect HTTP cache headers.
+ *
+ * This splits the header and body Promises so that things only needing
+ * headers don't need to wait on the body.
+ * @param {URL} parsed
+ * @param {ESModuleContext} context
+ * @returns {ReturnType<typeof fetchWithRedirects>}
+ */
+function fetch(parsed, { parentURL }) {
+  const { href } = parsed;
+  const existing = cacheForGET.get(href);
+  if (existing) {
+    return existing;
+  }
+  if (parsed.protocol === 'http:') {
+    return isLocalAddress(parsed.hostname).then((is) => {
```

Comment:
```suggestion
    return PromisePrototypeThen(isLocalAddress(parsed.hostname), (is) => {
```
or, make this an `async function` and await it.